From 7c25ea77c81e63c88cf1fbeb2fc9baba94bce8b7 Mon Sep 17 00:00:00 2001
From: Gary Ching-Pang Lin <glin@suse.com>
Date: Mon, 4 Mar 2013 16:25:08 +0800
Subject: [PATCH 4/9] Include the issuer's certificate only when available

When pesign generates a signature, it also includes the issuer's certificate.
In SUSE build server, we only import the signer's certificate and pesign
complaint the issuer's certificate was not found. Per Authenticode PE, the
root certificate is typically not included in the certificate list, so I
modified pesign a bit to include the issuer's certificate only when available.
Please check the attached patch.

Besides the issuer's certificate, I also found find_named_certificate() didn't
handle the certificate list properly and it may cause segfault if "node->cert"
is not valid. The patch also fixes this issue.
---
 src/cms_common.c  | 2 +-
 src/signed_data.c | 8 ++------
 2 files changed, 3 insertions(+), 7 deletions(-)

diff --git a/src/cms_common.c b/src/cms_common.c
index 6b44024..fc9796e 100644
--- a/src/cms_common.c
+++ b/src/cms_common.c
@@ -592,7 +592,7 @@ find_named_certificate(cms_context *cms, char *name, CERTCertificate **cert)
 	 * in the database, we'll get back what is essentially a template
 	 * that's in NSS's cache waiting to be filled out.  We can't use that,
 	 * it'll just cause CERT_DupCertificate() to segfault. */
-	if (!node || !node->cert || !node->cert->derCert.data
+	if (CERT_LIST_END(node) || !node->cert || !node->cert->derCert.data
 				 || !node->cert->derCert.len
 				 || !node->cert->derIssuer.data
 				 || !node->cert->derIssuer.len) {
diff --git a/src/signed_data.c b/src/signed_data.c
index 5425271..2f4b498 100644
--- a/src/signed_data.c
+++ b/src/signed_data.c
@@ -96,12 +96,8 @@ generate_certificate_list(cms_context *cms, SECItem ***certificate_list_p)
 		CERTCertificate *signer = NULL;
 		int rc = find_named_certificate(cms, cms->cert->issuerName,
 						&signer);
-		if (rc < 0) {
-			PORT_ArenaRelease(cms->arena, mark);
-			return -1;
-		}
-
-		if (signer && signer->derCert.len && signer->derCert.data) {
+		if (rc == 0 && signer &&
+				signer->derCert.len && signer->derCert.data) {
 			if (signer->derCert.len != cms->cert->derCert.len ||
 					memcmp(signer->derCert.data,
 						cms->cert->derCert.data,
-- 
1.8.3.1

